client-index/src/auto_synchronizer.rs

```diff
@@ -0,0 +1,746 @@
+            my_sender: None,
+            /// websocket receiver
+            my_receiver: None,
+        }
+    }
+
+    /// launch core thread
+    pub fn run<S: Storage + 'static, C: Client + 'static, H: BlockHandler + 'static>(
+        &mut self,
+        client: C,
+        storage: S,
+        block_handler: H,
+        data: AutoSyncDataShared,
+    ) {
+        let channel = futures::sync::mpsc::channel(0);
+        // tx, rx
+        let (channel_tx, channel_rx) = channel;
+        self.my_sender = Some(channel_tx.clone());
+        self.my_receiver = Some(channel_rx);
+
+        let mut core = AutoSynchronizerCore::new(
+            channel_tx.clone(),
+            storage,
+            client,
+            block_handler,
+            vec![],
+            data,
+        );
+        // save send_queue to communicate with core
+        self.core = Some(core.get_queue());
+
+        let _child = thread::spawn(move || {
+            core.start();
+        });
+
+        assert!(self.core.is_some());
+    }
+
+    /// activate tokio websocket
+    pub fn run_network(&mut self) -> Result<()> {
+        log::info!("Connecting to {}", self.websocket_url);
+        let mut runtime = tokio::runtime::current_thread::Builder::new()
+            .build()
+            .expect("get tokio builder");
+        // get synchronous sink
+        assert!(self.my_sender.is_some());
+        assert!(self.my_receiver.is_some());
+        let channel_tx = self.my_sender.as_ref().expect("get ref").clone();
+        let channel_rx = self.my_receiver.take().expect("take");
+        let mut channel_sink = channel_tx.clone().wait();
+
+        let runner = ClientBuilder::new(&self.websocket_url)
+            .expect("client-builder new")
+            .add_protocol("rust-websocket")
+            .async_connect_insecure()
+            .and_then(|(duplex, _)| {
+                channel_sink
+                    .send(OwnedMessage::Text(CMD_SUBSCRIBE.to_string()))
+                    .expect("send to channel sink");
+                let (sink, stream) = duplex.split();
+
+                stream
+                    .filter_map(|message| match message {
+                        OwnedMessage::Text(a) => {
+                            if let Some(core) = self.core.as_ref() {
+                                core.send(OwnedMessage::Text(a.clone())).expect("core send");
+                            }
+
+                            None
+                        }
+                        OwnedMessage::Binary(_a) => None,
+                        OwnedMessage::Close(e) => Some(OwnedMessage::Close(e)),
+                        OwnedMessage::Ping(d) => Some(OwnedMessage::Pong(d)),
+                        _ => None,
+                    })
+                    .select(channel_rx.map_err(|_| WebSocketError::NoDataAvailable))
+                    .forward(sink)
+            });
+        let _ = runtime.block_on(runner).expect("tokio block_on");
+        Ok(())
+    }
+}
```

Comment:
this module has over 700lines and seems to mash up a few things together -- it could make it a bit more readable to break it down into a few smaller logical units (commands, core synchroniser...)

Reply:
ok, i will divide into a few files.